packages/context/src/__tests__/unit/handler-chain.unit.ts

```diff
@@ -0,0 +1,168 @@
+  Next,
+} from '../..';
+
+describe('InvocationHandlerChain', () => {
+  let ctx: Context;
+  let handlerChain: InvocationHandlerChain;
+  let events: string[];
+
+  beforeEach(givenContext);
+
+  it('invokes handler functions', async () => {
+    givenHandlerChain(
+      givenNamedHandler('handler1'),
+      givenNamedHandler('handler2'),
+    );
+    const result = await handlerChain.invokeHandlers();
+    expect(events).to.eql([
+      'before-handler1',
+      'before-handler2',
+      'after-handler2',
+      'after-handler1',
+    ]);
+    expect(result).to.be.undefined();
+  });
+
+  it('honors return value', async () => {
+    givenHandlerChain(givenNamedHandler('handler1'), async (context, next) => {
+      await next();
+      return 'ABC';
+    });
+    const result = await handlerChain.invokeHandlers();
+    expect(result).to.eql('ABC');
+  });
+
+  it('skips downstream handlers if next is not invoked', async () => {
+    givenHandlerChain(async (context, next) => {
+      return 'ABC';
+    }, givenNamedHandler('handler1'));
+    await handlerChain.invokeHandlers();
+    expect(events).to.eql([]);
+  });
+
+  it('passes bindings via context', async () => {
+    givenHandlerChain(
+      async (context, next) => {
+        context.bind('foo').to('1-req');
+        await next();
+        const foo = await context.get('foo');
+        expect(foo).to.eql('2-res');
+        context.bind('foo').to('1-res');
+      },
+      async (context, next) => {
+        const foo = await context.get('foo');
+        expect(foo).to.eql('1-req');
+        await next();
+        context.bind('foo').to('2-res');
+      },
+    );
+
+    await handlerChain.invokeHandlers();
+    const fooVal = await ctx.get('foo');
+    expect(fooVal).to.eql('1-res');
+  });
+
+  it('catches error from second handler', async () => {
+    givenHandlerChain(givenNamedHandler('handler1'), async (context, next) => {
+      events.push('before-handler2');
+      throw new Error('error in handler2');
+    });
+    const resultPromise = handlerChain.invokeHandlers();
+    await expect(resultPromise).to.be.rejectedWith('error in handler2');
+    expect(events).to.eql(['before-handler1', 'before-handler2']);
+  });
+
+  it('catches error from first handler', async () => {
+    givenHandlerChain(async (context, next) => {
+      events.push('before-handler1');
+      await next();
+      throw new Error('error in handler1');
+    }, givenNamedHandler('handler2'));
+    const resultPromise = handlerChain.invokeHandlers();
+    await expect(resultPromise).to.be.rejectedWith('error in handler1');
+    expect(events).to.eql([
+      'before-handler1',
+      'before-handler2',
+      'after-handler2',
+    ]);
+  });
+
+  it('allows discovery of handlers in context', async () => {
+    const handler1 = givenNamedHandler('handler1');
+    const handler2 = givenNamedHandler('handler2');
+    ctx
+      .bind('handler2')
+      .to(handler2)
+      .tag('handler');
```

Comment:
Should we have a binding template `asHandler` to set the tag and possibly other binding metadata as needed? Unless the tag `handler` is arbitrary and used by the test only, in which case please make that more obvious by using a different tag name, e.g. `some-tag`.